### ISSUE TITLE
Tab titles always show title from first tab

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#778]
This was reported by several people on Gentoo and one on Archlinux - but I can't reproduce it so far.

It seems for those, after a recent (Qt/qutebrowser?) update, all tabs show the title from the first one:

![qtbug](https://cloud.githubusercontent.com/assets/625793/8323112/4ed348cc-1a40-11e5-8753-36dbd381829a.png)

When closing the leftmost tab, all tabs get the title from the next one. It seems it also happens with the default config (`-c ""`).

In the logs, things look like expected:

```
webview    tabbedbrowser:on_title_changed:455 Changing title for idx 2 to 'http://www.spiegel.de/auto/aktuell/nautilus-car-sechsraedriges-filmauto-wird-versteigert-a-1039657.html'
webview    tabbedbrowser:on_title_changed:447 Ignoring title change to ''.
webview    tabbedbrowser:on_title_changed:455 Changing title for idx 2 to 'Nautilus Car: Sechsrädriges Filmauto wird versteigert - SPIEGEL ONLINE'
```

/cc @rieper @vekin (I hope these are the right people - sorry if not!)

**Comment by The-Compiler** [comment]
Downgrading from Qt 5.5 to 5.4.2 reportedly fixes the issue on Archlinux... :worried:

**Comment by ff2000** [comment]
Florian Bruhin wrote:
> Downgrading from Qt 5.5 to 5.4.2 reportedly fixes the issue on Archlinux... :worried: 

Gentoo here. Just tested and it seems to happen here, too.
But still on 5.4.2... So no chance to downgrade ;)

**Comment by The-Compiler** [comment]
@ff2000 yeah, I've heard about several Gentoo people (with 5.4.2) with the issue as well... Yet looking at [the package](https://packages.gentoo.org/package/dev-qt/qtwidgets) I don't see any Gentoo-specific patches which could cause it

Can you please reproduce it on the `issue778` branch and start with `--debug --logfilter title` and post the log here (or do `:report`)?

**Comment by ff2000** [comment]
Sent with :report.
Would be great if the dialog could tell what actually will happen: Create a public issue? Or send you a private Mail? For the last one I would allow certain private details in the log ("trust" ;))

**Comment by The-Compiler** [comment]
Thanks! So when you did `:report`, both tabs had `phaeozem bei DuckDuckGo` as title?

It seems like the title and tab data is set correctly:

```
12:39:27 DEBUG    title      tabwidget:update_tab_title:124 Updating page title for 0: phaeozem bei DuckDuckGo
[...]
12:39:29 DEBUG    title      tabwidget:update_tab_title:124 Updating page title for 1: Phaeozem – Wikipedia
[...]
12:39:29 DEBUG    title      tabwidget:tab_data:283 Getting tab data (key indicator-color) for 0 -> {'page-title': 'phaeozem bei DuckDuckGo', 'indicator-color': ...}
12:39:29 DEBUG    title      tabwidget:tab_data:283 Getting tab data (key indicator-color) for 1 -> {'page-title': 'Phaeozem – Wikipedia', 'indicator-color': ...}
```

Can you try running this minimal example please?

```python
from PyQt5.QtWidgets import QApplication, QTabBar

app = QApplication([])
bar = QTabBar()

bar.addTab("foo")
bar.addTab("bar")

bar.show()
app.exec_()
```

Do you see a `foo` and `bar` tab there?

The crash reports are public [in my pastebin](http://p.cmpl.cc/) currently. I thought the report dialog said that, but it seems it doesn't... I want to change then to be private anyways though - see #481. (And now I see I already said the same regarding the report dialog there... Too many things I want to take a look at!)

**Comment by ff2000** [comment]
Just a hint: The issue is in TabBarStyle.
When I comment out the line that sets the style in TabBar this issue disappears.
And you might want to check QProxyStyle, which AFAIK was designed to proxy the style ;) (Though I might be wrong here, didn't dive too deep into the code)

**Comment by The-Compiler** [comment]
Thanks, I'll take a closer look at that later. Unfortunately `QProxyStyle` isn't available in PyQt - that would make things a lot easier indeed!

**Comment by The-Compiler** [comment]
I've now tried again on my system after installing the `-beta` packages again, i.e. the Qt 5.5 RC. Unfortunately I still can't reproduce it... :worried:

What desktop environment are you using, if any? Does starting with `--qt-style Fusion` or `--qt-style Windows` change anything?

**Comment by kirelagin** [comment]
Yep, same issue. Gentoo, Qt 5.4.2. No DE, just xmonad.
Neither `--qt-style` fixed this. I sent you a `:report`.

**Comment by ff2000** [comment]
Well...

```
diff --git a/qutebrowser/mainwindow/tabwidget.py b/qutebrowser/mainwindow/tabwidget.py
index 0f70a44..297a5e8 100644
--- a/qutebrowser/mainwindow/tabwidget.py
+++ b/qutebrowser/mainwindow/tabwidget.py
@@ -401,9 +401,9 @@ class TabBar(QTabBar):
     def paintEvent(self, _e):
         """Override paintEvent to draw the tabs like we want to."""
         p = QStylePainter(self)
-        tab = QStyleOptionTab()
         selected = self.currentIndex()
         for idx in range(self.count()):
+            tab = QStyleOptionTab()
             self.initStyleOption(tab, idx)
             if idx == selected:
                 bg_color = config.get('colors', 'tabs.bg.selected')
```
Remember: Keep variables as local as possible...
That fixes the issue for me. Please test @kirelagin (and others)
My python is too bad, but I can imagine it has to do with strings not being assigned/unreferenced/garbage collected properly, so more a python/sip issue than one in Qt.

**Comment by The-Compiler** [comment]
Whoa, good catch!

Instead of using a "clean" `QStyleOptionTab` object every time, the same is reused. And I guess `initStyleOption` doesn't set the text correctly in some circumstances - as you said.

I'm going to assume this fixes it for everyone - if not, please let me know and I'll reopen.

I was just downloading a Gentoo Live CD trying to reproduce this - so thanks a lot to @ff2000 for saving me a lot of time and for the fix! :+1:

**Comment by ff2000** [comment]
@The-Compiler which sip-version are you using (here: 4.16.8)? I can imagine the bug is there (marshalling python strings<->QStrings).
Concerning Gentoo: You would have to actually install it. Qt5 isn't stable yet, so you have to learn how to install masked packages. And the ISOs only contain stable software...

**Comment by kirelagin** [comment]
Yep, that fixed it :+1:.

**Comment by The-Compiler** [comment]
@ff2000 I'm on sip 4.16.8 as well. Qt5 is still unstable after 2.5 years, and less than half a year before Qt4 is unsupported upstream? Whoa...

@kirelagin Glad to hear!

**Comment by ff2000** [comment]
Florian Bruhin wrote:
> @ff2000 I'm on sip 4.16.8 as well. Qt5 is still unstable after 2.5 years, and less than half a year before Qt4 is unsupported upstream? Whoa...

Here is the Gentoo qt-5.4.2 stable tracker:
https://bugs.gentoo.org/show_bug.cgi?id=543326

There are still open issues.
Also some of the fixed bugs took some time to get fixed.
Qt5 will hit Gentoo-stable when the Gentoo Qt-herd considers it stable ;) And it will most likely be before Qt4 hits "unmaintained" :)
